// @liveblocks/react-tiptap/dist/index.cjs
'use strict';

var core = require('@liveblocks/core');
var version = require('./version.cjs');
var AiToolbar = require('./ai/AiToolbar.cjs');
var AnchoredThreads = require('./comments/AnchoredThreads.cjs');
var FloatingComposer = require('./comments/FloatingComposer.cjs');
var FloatingThreads = require('./comments/FloatingThreads.cjs');
var LiveblocksExtension = require('./LiveblocksExtension.cjs');
var GroupMentionNode = require('./mentions/GroupMentionNode.cjs');
var MentionExtension = require('./mentions/MentionExtension.cjs');
var MentionNode = require('./mentions/MentionNode.cjs');
var FloatingToolbar = require('./toolbar/FloatingToolbar.cjs');
var Toolbar = require('./toolbar/Toolbar.cjs');
var HistoryVersionPreview = require('./version-history/HistoryVersionPreview.cjs');

core.detectDupes(version.PKG_NAME, version.PKG_VERSION, version.PKG_FORMAT);

exports.AiToolbar = AiToolbar.AiToolbar;
exports.AnchoredThreads = AnchoredThreads.AnchoredThreads;
exports.FloatingComposer = FloatingComposer.FloatingComposer;
exports.FloatingThreads = FloatingThreads.FloatingThreads;
exports.useIsEditorReady = LiveblocksExtension.useIsEditorReady;
exports.useLiveblocksExtension = LiveblocksExtension.useLiveblocksExtension;
exports.GroupMentionNode = GroupMentionNode.GroupMentionNode;
exports.MentionExtension = MentionExtension.MentionExtension;
exports.MentionNode = MentionNode.MentionNode;
exports.FloatingToolbar = FloatingToolbar.FloatingToolbar;
exports.Toolbar = Toolbar.Toolbar;
exports.HistoryVersionPreview = HistoryVersionPreview.HistoryVersionPreview;
