// @liveblocks/react-tiptap/dist/types.cjs
'use strict';

var state = require('@tiptap/pm/state');

const LIVEBLOCKS_MENTION_KEY = new state.PluginKey("lb-plugin-mention");
const LIVEBLOCKS_MENTION_PASTE_KEY = new state.PluginKey(
  "lb-plugin-mention-paste"
);
const LIVEBLOCKS_MENTION_NOTIFIER_KEY = new state.PluginKey(
  "lb-plugin-mention-notify"
);
const LIVEBLOCKS_MENTION_EXTENSION = "liveblocksMentionExt";
const LIVEBLOCKS_MENTION_TYPE = "liveblocksMention";
const LIVEBLOCKS_GROUP_MENTION_TYPE = "liveblocksGroupMention";
const THREADS_ACTIVE_SELECTION_PLUGIN = new state.PluginKey(
  "lb-threads-active-selection-plugin"
);
const THREADS_PLUGIN_KEY = new state.PluginKey(
  "lb-threads-plugin"
);
const AI_TOOLBAR_SELECTION_PLUGIN = new state.PluginKey(
  "lb-ai-toolbar-selection-plugin"
);
const LIVEBLOCKS_COMMENT_MARK_TYPE = "liveblocksCommentMark";
var ThreadPluginActions = /* @__PURE__ */ ((ThreadPluginActions2) => {
  ThreadPluginActions2["SET_SELECTED_THREAD_ID"] = "SET_SELECTED_THREAD_ID";
  return ThreadPluginActions2;
})(ThreadPluginActions || {});

exports.AI_TOOLBAR_SELECTION_PLUGIN = AI_TOOLBAR_SELECTION_PLUGIN;
exports.LIVEBLOCKS_COMMENT_MARK_TYPE = LIVEBLOCKS_COMMENT_MARK_TYPE;
exports.LIVEBLOCKS_GROUP_MENTION_TYPE = LIVEBLOCKS_GROUP_MENTION_TYPE;
exports.LIVEBLOCKS_MENTION_EXTENSION = LIVEBLOCKS_MENTION_EXTENSION;
exports.LIVEBLOCKS_MENTION_KEY = LIVEBLOCKS_MENTION_KEY;
exports.LIVEBLOCKS_MENTION_NOTIFIER_KEY = LIVEBLOCKS_MENTION_NOTIFIER_KEY;
exports.LIVEBLOCKS_MENTION_PASTE_KEY = LIVEBLOCKS_MENTION_PASTE_KEY;
exports.LIVEBLOCKS_MENTION_TYPE = LIVEBLOCKS_MENTION_TYPE;
exports.THREADS_ACTIVE_SELECTION_PLUGIN = THREADS_ACTIVE_SELECTION_PLUGIN;
exports.THREADS_PLUGIN_KEY = THREADS_PLUGIN_KEY;
exports.ThreadPluginActions = ThreadPluginActions;
