// @liveblocks/react-ui/dist/primitives/slate/utils/marks.js
import { Editor, Range } from 'slate';
import { getCharacterBefore, getCharacterAfter } from './get-character.js';

const defaultComposerBodyMarks = {
  bold: false,
  italic: false,
  strikethrough: false,
  code: false
};
function getComposerBodyMarks(editor) {
  if (!editor) {
    return { ...defaultComposerBodyMarks };
  }
  const marks = Editor.marks(editor);
  return { ...defaultComposerBodyMarks, ...marks };
}
function isMarkActive(editor, mark) {
  const marks = Editor.marks(editor);
  return marks ? marks[mark] === true : false;
}
function filterActiveMarks(value) {
  return Object.keys(value ?? {}).filter(
    (key) => key !== "text"
  );
}
function toggleMark(editor, mark) {
  const isActive = isMarkActive(editor, mark);
  if (isActive) {
    Editor.removeMark(editor, mark);
  } else {
    Editor.addMark(editor, mark, true);
  }
}
function removeMarks(editor) {
  const marks = Editor.marks(editor);
  if (marks) {
    for (const mark in marks) {
      Editor.removeMark(editor, mark);
    }
  }
}
function leaveMarkEdge(editor, edge) {
  if (editor.selection && Range.isCollapsed(editor.selection)) {
    const marks = Object.keys(Editor.marks(editor) ?? {});
    if (marks.length > 0) {
      const sibling = edge === "start" ? getCharacterBefore(editor, editor.selection) : getCharacterAfter(editor, editor.selection);
      if (!sibling) {
        removeMarks(editor);
      }
    }
  }
}

export { filterActiveMarks, getComposerBodyMarks, isMarkActive, leaveMarkEdge, removeMarks, toggleMark };
