// @liveblocks/react-ui/dist/components/internal/Prose.cjs
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var components = require('../../components.cjs');
var Markdown = require('../../primitives/Markdown.cjs');
var cn = require('../../utils/cn.cjs');
var CodeBlock$1 = require('./CodeBlock.cjs');

function Link({ href, title, children }) {
  const { Anchor } = components.useComponents();
  return /* @__PURE__ */ jsxRuntime.jsx(Anchor, {
    href,
    title,
    children
  });
}
function CodeBlock({ language, code }) {
  return /* @__PURE__ */ jsxRuntime.jsx(CodeBlock$1.CodeBlock, {
    title: language || "Plain text",
    code
  });
}
const defaultMarkdownComponents = {
  Link,
  CodeBlock
};
function Prose({
  content,
  partial,
  components: components$1,
  className,
  ...props
}) {
  return /* @__PURE__ */ jsxRuntime.jsx(components.ComponentsProvider, {
    components: components$1,
    children: /* @__PURE__ */ jsxRuntime.jsx(Markdown.Markdown, {
      content,
      components: { ...defaultMarkdownComponents, ...components$1?.markdown },
      className: cn.cn("lb-root lb-prose", className),
      partial,
      ...props
    })
  });
}

exports.Prose = Prose;
