// @liveblocks/react-tiptap/dist/mentions/MentionNode.cjs
'use strict';

var core = require('@tiptap/core');
var react = require('@tiptap/react');
var types = require('../types.cjs');
var Mention = require('./Mention.cjs');

const MentionNode = core.Node.create({
  name: types.LIVEBLOCKS_MENTION_TYPE,
  group: "inline",
  inline: true,
  selectable: true,
  atom: true,
  priority: 101,
  parseHTML() {
    return [
      {
        tag: "liveblocks-mention"
      }
    ];
  },
  renderHTML({ HTMLAttributes }) {
    return ["liveblocks-mention", core.mergeAttributes(HTMLAttributes)];
  },
  addNodeView() {
    return react.ReactNodeViewRenderer(Mention.Mention, {
      contentDOMElementTag: "span"
    });
  },
  addAttributes() {
    return {
      id: {
        default: null,
        parseHTML: (element) => element.getAttribute("data-id"),
        renderHTML: (attributes) => {
          if (!attributes.id) {
            return {};
          }
          return {
            "data-id": attributes.id
          };
        }
      },
      notificationId: {
        default: null,
        parseHTML: (element) => element.getAttribute("data-notification-id"),
        renderHTML: (attributes) => {
          if (!attributes.notificationId) {
            return {};
          }
          return {
            "data-notification-id": attributes.notificationId
          };
        }
      }
    };
  },
  addKeyboardShortcuts() {
    return {
      Backspace: () => this.editor.commands.command(({ tr, state }) => {
        let isMention = false;
        const { selection } = state;
        const { empty, anchor } = selection;
        if (!empty) {
          return false;
        }
        state.doc.nodesBetween(anchor - 1, anchor, (node, pos) => {
          if (node.type.name === this.name) {
            isMention = true;
            tr.insertText("", pos, pos + node.nodeSize);
          }
        });
        return isMention;
      })
    };
  }
});

exports.MentionNode = MentionNode;
