// @liveblocks/react-tiptap/dist/mentions/Mention.cjs
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var core = require('@liveblocks/core');
var _private = require('@liveblocks/react-ui/_private');
var react$1 = require('@tiptap/react');
var react = require('react');
var types = require('../types.cjs');

const UserMention = react.forwardRef(
  ({ mention, isSelected }, forwardedRef) => {
    return /* @__PURE__ */ jsxRuntime.jsxs(react$1.NodeViewWrapper, {
      className: _private.cn(
        "lb-root lb-mention lb-tiptap-mention",
        isSelected && "lb-mention-selected"
      ),
      as: "span",
      ref: forwardedRef,
      children: [
        /* @__PURE__ */ jsxRuntime.jsx("span", {
          className: "lb-mention-symbol",
          children: core.MENTION_CHARACTER
        }),
        /* @__PURE__ */ jsxRuntime.jsx(_private.User, {
          userId: mention.id
        })
      ]
    });
  }
);
const GroupMention = react.forwardRef(
  ({ mention, isSelected }, forwardedRef) => {
    return /* @__PURE__ */ jsxRuntime.jsxs(react$1.NodeViewWrapper, {
      className: _private.cn(
        "lb-root lb-mention lb-tiptap-mention",
        isSelected && "lb-mention-selected"
      ),
      as: "span",
      ref: forwardedRef,
      children: [
        /* @__PURE__ */ jsxRuntime.jsx("span", {
          className: "lb-mention-symbol",
          children: core.MENTION_CHARACTER
        }),
        /* @__PURE__ */ jsxRuntime.jsx(_private.Group, {
          groupId: mention.id
        })
      ]
    });
  }
);
function deserializeGroupUserIds(userIds) {
  if (typeof userIds !== "string") {
    return void 0;
  }
  try {
    const parsedUserIds = JSON.parse(userIds);
    if (Array.isArray(parsedUserIds)) {
      return parsedUserIds;
    }
    return void 0;
  } catch {
    return void 0;
  }
}
const Mention = react.forwardRef(({ node, selected: isSelected }, forwardedRef) => {
  const attrs = node.attrs;
  if (node.type.name === types.LIVEBLOCKS_MENTION_TYPE) {
    const mention = {
      kind: "user",
      id: attrs.id
    };
    return /* @__PURE__ */ jsxRuntime.jsx(UserMention, {
      mention,
      isSelected,
      ref: forwardedRef
    });
  }
  if (node.type.name === types.LIVEBLOCKS_GROUP_MENTION_TYPE) {
    const mention = {
      kind: "group",
      id: attrs.id,
      userIds: deserializeGroupUserIds(attrs.userIds)
    };
    return /* @__PURE__ */ jsxRuntime.jsx(GroupMention, {
      mention,
      isSelected,
      ref: forwardedRef
    });
  }
  return null;
});

exports.Mention = Mention;
